### ISSUE TITLE
[libc][math][c23] Implement C23 math functions fromfp, fromfpx, ufromfp, ufromfpx

### ISSUE THREAD

**Issue by lntue** [llvm/llvm-project#85279]
These functions are defined in C23 standard: section 7.12.9.10 and 7.12.9.11 in https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3096.pdf.

We will need to adapt our current rounding functions to extra requirements of these functions.

Need to support the following data types: float, double, long double, and float128.

**Comment by overmighty** [comment]
I am interested in implementing this.

**Comment by overmighty** [comment]
There's something I'm not sure I understand in the description of `fromfp` and `ufromfp` in section 7.12.9.10:

> [...] **If `width` is nonzero and the resulting integer is within the
> range**
> — [−2^(`width`−1), 2^(`width`−1) − 1], for signed
> — [0, 2^`width` − 1], for unsigned
> then the functions return the integer value (represented in floating type). Otherwise, **if `width` is
> zero or `x` does not round to an integer within the range**, the functions return a NaN (of the type
> of the `x` argument, if available), **else** the value of `x`, and a domain error occurs. [...]

The second emphasized "if" is the negation of the first, so already an else. Does the following emphasized "else" mean "if a NaN of the type of the `x` argument is not available"? Maybe the parentheses shouldn't be there.

**Comment by lntue** [comment]
> There's something I'm not sure I understand in the description of `fromfp` and `ufromfp` in section 7.12.9.10:
> 
> > [...] **If `width` is nonzero and the resulting integer is within the
> > range**
> > — [−2^(`width`−1), 2^(`width`−1) − 1], for signed
> > — [0, 2^`width` − 1], for unsigned
> > then the functions return the integer value (represented in floating type). Otherwise, **if `width` is
> > zero or `x` does not round to an integer within the range**, the functions return a NaN (of the type
> > of the `x` argument, if available), **else** the value of `x`, and a domain error occurs. [...]
> 
> The second emphasized "if" is the negation of the first, so already an else. Does the following emphasized "else" mean "if a NaN of the type of the `x` argument is not available"? Maybe the parentheses shouldn't be there.

You are right, logically the sentence in the standard does not make sense.  Probably the correct one is:
> Otherwise, if `width` is zero or `x` does not round to an integer within the range, the functions return
> (a NaN of the type of the `x` argument **if** available, **else** the value of `x`, and a domain error occurs ). [...]

@AaronBallman : is that the correct intention of the C23 standard?

**Comment by AaronBallman** [comment]
I reached out to a member of the C Floating Point study group just to double-check our understanding, and yes, this is a bug in the standard. The way it should be interpreted is:

Otherwise, if width is zero or x does not round to an integer within the range
  * the functions return a NaN (of the type of the x argument), if available,
  * else the value of x,
and a domain error occurs.

We'll work on getting the standard fixed up, thank you for raising the question!
